### DISCUSSION TITLE
Using Apollo federation directives

### DISCUSSION THREAD

**Discussion by shamin2021** [graph-quilt/graphql-orchestrator-java#191] [General]
Can you please let me know how to stitch when I have schemas like below, Since i am getting an error in compilation 

    public static final String schema =
                    "type TaxInfo @extends @key(fields: 'taxId year') { taxId: ID! @external year: Int! @external taxDocument:TaxDocument}  " +
                    "type TaxDocument { year: Int forms: [TaxForm] }" +
                    "type TaxForm { new : String } ";
                    
    public static final String schema =
                    "type Query { user(id: ID!): User } " +
                    "type User { finance: FinanceInfo }  " +
                    "type FinanceInfo { tax(year: Int!): TaxInfo! } " +
                    " type TaxInfo @key(fields: 'taxId year') { taxId: ID! year: Int! } ";
                    
                    
 ERROR:Couldn't resolve reference to DirectiveDefinition 'key'. (schema.graphqls line : 1 column : 134) 
 These parsing issues were found in the following files.

**Comment by ashpak-shaikh** [comment]
Hey @shamin2021, 

Thanks for reaching out. 

You need to set the type as [FEDERATION_SUBGRAPH](https://github.com/graph-quilt/graphql-orchestrator-java/blob/master/src/main/java/com/intuit/graphql/orchestrator/ServiceProvider.java#L40) to activate the federation style stitching. Please try overriding the getServiceType() method in ServiceProvider interface

**Reply by shamin2021** [reply]
can you please let me know how to implement datafetchers for each subgraph as well

**Comment by shamin2021** [comment]
//ProductName Service 
    @Override
    public CompletableFuture<Map<String, Object>> query(final ExecutionInput executionInput,
                                                        final GraphQLContext context) {
        Map<String, Object> data = ImmutableMap
                .of("data", ImmutableMap.of("product", ImmutableMap.of("id", 1, "name", "Java","description","hjsaj")));
        return CompletableFuture.completedFuture(data);
    }
    
        //ProductReviewsService 
        @Override
    public CompletableFuture<Map<String, Object>> query(final ExecutionInput executionInput,
                                                        final GraphQLContext context) {

        Map<String, Object> data = ImmutableMap
                .of("data", ImmutableMap.of("product", ImmutableMap.of("reviews", ImmutableMap.of("id",2 , "text","CA", "starRating",2))));
        return CompletableFuture.completedFuture(data);

    }
 

Schema 1 
    type Product @key(fields: "id") @extends {
    id: ID! @external
    reviews: [Review!]!
}

type Review {
    id: ID!,
    text: String
    starRating: Int!
}

Schema 2
type Query {
    product(id: ID!): Product
    products: [Product!]!
}

type Product @key(fields: "id") {
    id: ID!
    name: String!
    description: String
}

I want to run query query { product (id : 1){ reviews { text } } } please let me know how i can Achieve this 

I am getting the following error
ExecutionResultImpl{errors=[ExceptionWhileDataFetching{path=[product], exception=java.lang.IllegalAccessError: class com.intuit.graphql.orchestrator.common.ArgumentValueResolver tried to access private method 'void graphql.execution.ValuesResolver.<init>()' (com.intuit.graphql.orchestrator.common.ArgumentValueResolver and graphql.execution.ValuesResolver are in unnamed module of loader 'app'), locations=[SourceLocation{line=1, column=9}]}], data={product=null}, dataPresent=true, extensions=null}

**Reply by ashpak-shaikh** [reply]
https://netflix.github.io/dgs/federation/

This is a good example of how to write a Data Fetcher for federation, basically u need an entity data fetcher. The call it will make to the review service is an entity fetch call using the representation query that need to be resolved using an entity data fetcher. An entity data fetcher is not a regular data fetcher, but a special datafetcher in the federation world

https://github.com/graph-quilt/graphql-orchestrator-java/blob/master/src/test/groovy/com/intuit/graphql/orchestrator/integration/federation/EntityTypeSpec.groovy this test has some federation examples with mock responses as well.

**Reply by shamin2021** [reply]
To clear some things up, DGS framework does not provide a  method for composing a supergraph right?

**Reply by shamin2021** [reply]
and how do we define endpoints in the orchestrator approach (ex: const gateway = new ApolloGateway({
    supergraphSdl: new IntrospectAndCompose({
        subgraphs: [
            { name: 'shows', url: 'http://localhost:8080/graphql' },
            { name: 'reviews', url: 'http://localhost:8081/graphql' },
        ]
    })
});)

**Comment by ashpak-shaikh** [comment]
If you are planning to use this as a gateway, please use https://github.com/graph-quilt/graphql-gateway-java as the reference implementation. It has a few subgraphs in the local-registry folder, you modify the schema and play around with it.

**Reply by ashpak-shaikh** [reply]
we can also setup some time aside to speed things up if you like.

**Comment by ashpak-shaikh** [comment]
```
class ProductNameService implements ServiceProvider {

    String filePath1 = "src/main/resources/Product.graphql";

    public String schema;
    {
        try {
            schema = readFileAsString(filePath1);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }


    @Override
    public String getNameSpace() { return "PRODUCT"; }

    @Override
    public ServiceType getSeviceType() {
        return ServiceType.FEDERATION_SUBGRAPH;
    }

    @Override
    public Map<String, String> sdlFiles() { return ImmutableMap.of("schema.graphqls", schema); }

    @Override
    public CompletableFuture<Map<String, Object>> query(final ExecutionInput executionInput,
                                                        final GraphQLContext context) {
        Map<String, Object> data = ImmutableMap
            .of("data", ImmutableMap.of("getBar", ImmutableMap.of("id", "bar")));
        System.out.println("PRODUCT_QUERY");

        System.out.println(data);
        return CompletableFuture.completedFuture(data);
        //return null;
    }

    public static String readFileAsString(String filePath) throws Exception {
        Path path = Paths.get(filePath);
        byte[] bytes = Files.readAllBytes(path);
        String content = new String(bytes, StandardCharsets.UTF_8);
        return content;
    }
}
```

```
class ProductReviewService implements ServiceProvider {

    String filePath1 = "src/main/resources/Reviews.graphql";
   // String filePath2 = "src/main/resources/resolverSchema.graphql";

    public  final String schema;
   // public  final String resolverSchema;

    {
        try {
            schema = readFileAsString(filePath1);
           // resolverSchema = readFileAsString(filePath2);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }


    @Override
    public String getNameSpace() { return "REVIEW";}

    @Override
    public Map<String, String> sdlFiles() { return ImmutableMap.of("schema.graphqls", schema);}

    @Override
    public ServiceType getSeviceType() {
        return ServiceType.FEDERATION_SUBGRAPH;
    }

    @Override
    public CompletableFuture<Map<String, Object>> query (final ExecutionInput executionInput, final GraphQLContext context) {
        Map<String, Object> data = ImmutableMap
                .of("data", ImmutableMap.of("getBar", ImmutableMap.of("id", "bar")));
        System.out.println("REVIEW_QUERY");

        System.out.println(data);
        return CompletableFuture.completedFuture(data);
       // return null;

    }

    public static String readFileAsString(String filePath) throws Exception {
        Path path = Paths.get(filePath);
        byte[] bytes = Files.readAllBytes(path);
        String content = new String(bytes, StandardCharsets.UTF_8);
        return content;
    }
```

pom dependencies are the problem in your case
```
  <dependencies>
        <dependency>
            <groupId>org.apache.logging.log4j</groupId>
            <artifactId>log4j-core</artifactId>
            <version>2.20.0</version>
        </dependency>

        <dependency>
            <groupId>org.apache.logging.log4j</groupId>
            <artifactId>log4j-slf4j2-impl</artifactId>
            <version>2.20.0</version>
        </dependency>

        <dependency>
            <groupId>org.slf4j</groupId>
            <artifactId>slf4j-api</artifactId>
            <version>2.0.7</version>
            <scope>runtime</scope>
        </dependency>
        <dependency>
            <groupId>ch.qos.logback</groupId>
            <artifactId>logback-classic</artifactId>
            <version>1.2.6</version> <!-- Use the latest version -->
        </dependency>
        <dependency>
            <groupId>org.apache.xbean</groupId>
            <artifactId>xbean-reflect</artifactId>
            <version>3.7</version>
        </dependency>
        <dependency>
            <groupId>com.fasterxml.jackson.core</groupId>
            <artifactId>jackson-databind</artifactId>
            <version>2.16.0</version>
        </dependency>

        <dependency>
            <groupId>com.intuit.graphql</groupId>
            <artifactId>graphql-orchestrator-java</artifactId>
            <version>LATEST</version>
        </dependency>
    </dependencies>
```

**Comment by shamin2021** [comment]
Thankyou so much this helped

**Comment by shamin2021** [comment]
Do we need to add a resolver schema as well ? to query two services? for an example 
query {
  getFoo {
    id
    fooName
    bar {
      id
      barName
    }
  }
}
